plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/HudiSplitSource.java

```diff
@@ -113,20 +133,39 @@ public HudiSplitSource(
                     queue.finish();
                 });
         this.splitLoaderFuture = splitLoaderExecutorService.schedule(splitLoader, 0, TimeUnit.MILLISECONDS);
+        this.dynamicFilter = requireNonNull(dynamicFilter, "dynamicFilter is null");
+        this.dynamicFilteringWaitTimeoutMillis = dynamicFilteringWaitTimeoutMillis.toMillis();
+        this.dynamicFilterWaitStopwatch = Stopwatch.createStarted();
     }
 
     @Override
     public CompletableFuture<ConnectorSplitBatch> getNextBatch(int maxSize)
     {
+        long timeLeft = dynamicFilteringWaitTimeoutMillis - dynamicFilterWaitStopwatch.elapsed(MILLISECONDS);
+        if (dynamicFilter.isAwaitable() && timeLeft > 0) {
+            return dynamicFilter.isBlocked()
+                    .thenApply(ignored -> EMPTY_BATCH)
+                    .completeOnTimeout(EMPTY_BATCH, timeLeft, MILLISECONDS);
```

Comment:
So does the logic give back the connector splits from the Trino query plan?  How does it work?

Reply:
Do the tests cover both logic, i.e., `if` branch and the logic below where `HudiSplitSource` uses the dynamic filter predicate for pruning?

Reply:
See:

1. testDynamicFilterEnabledPredicatePushdown
2. testDynamicFilterDisabledPredicatePushdown
3. testDynamicFilterEnabled__withPartitionPruningUsingDynamicFilter
4. testDynamicFilterDisabled__withPartitionPruningUsingDynamicFilter

Reply:
> So, when the probe side is invoking getNextBatch during the query execution stage, this logic delays the probe side until build finalizes the dynamicFilter by a configurable timeout.

> If it is not ready, an EmptyBatch CompletableFuture is returned. The calling function will have knowledge of whether there are more splits to be processed via #isFinished(). At this point, #getNextBatch() returns an EmptyBatch and #isFinished() is still false. So, it will invoke the #getNextBatch() concrete implementation again.

Got it, this is really helpful.  Could add a couple of lines of comments in `HudiSplitSource` on this so it's easier to understand the code in the future?

Reply:
So does this mean that the `HudiSplitSource` only waits for 1s at most for the dynamic filter to be ready?  Even if later on the dynamic filter is available during split generation, the filter may not be leveraged?